### PR TITLE
zos,test: free system resources on process exit

### PR DESCRIPTION
**Pull request by jBarz**
The libuv tests are leaving behind semaphore and message queues which persist on z/OS after process termination if not cleaned up. This code change ensures that the tests clean up on exit and
on abornal termination via signals.

This is to address the problems being seen on the z/OS test machine which runs out of system semaphores/messagequeues after a few test runs.

### CODE REVIEW

**Comment by gireeshpunathil** [comment]
- the cleanup routine `release_system_resources` - looks like it is added to a test file, not in the mainstream source, am I missing something?

 - who creates message queues and semaphores in the process? if they are through libuv wrappers, can we book-keep their handles every time one is created, and then delete them on exit, and enforce manual clean-up of these resources from user-land code, through a documentation? just a thought.

**Comment by jBarz** [comment]
> and enforce manual clean-up of these resources from user-land code

The way to do this would be to close the loop. A libuv user could use `uv_loop_close`. The test cases do this via `MAKE_VALGRIND_HAPPY` which is run at the end of each test case.

> the cleanup routine release_system_resources - looks like it is added to a test file, not in the mainstream source, am I missing something?

This affects the test cases. Any source code using libuv would have to do the same.
I added this signal handler which will query the OS for all the semaphores/msg queues that the current process has created and then remove them. I could simply close the event loop in the signal handler but wasn't sure if it is safe to do that.

**Comment by jBarz** [comment]
Another approach to this could be moving the release_system_resources into the `UV_DESTRUCTOR(static void cleanup(void))` function in threadpool.c.
This requires libuv to keep track of all semaphores/msgqueues it creates (for z/OS only).
And ensures that libuv always frees up system resources on exit even when the libuv user doesn't.

File: test/runner-unix.c [hunk 1]
```diff
@@ -53,6 +99,13 @@ int platform_init(int argc, char **argv) {
     return -1;
   }
 
+#ifdef __MVS__
+  memset(&action, 0, sizeof(struct sigaction));
```
Comment (bnoordhuis) [inline test/runner-unix.c]: `sizeof(action)`

File: test/task.h [hunk 2]
```diff
@@ -114,6 +114,12 @@ typedef enum {
     ASSERT(0 == uv_loop_close(uv_default_loop()));  \
   } while (0)
 
+#define MAKE_VALGRIND_HAPPY_1(loop)    \
+  do {                                 \
+    close_loop(loop);                  \
+    ASSERT(0 == uv_loop_close(loop));  \
+  } while (0)
```
Comment (bnoordhuis) [inline test/task.h]: Debug remnant?  It's only used in one place.

**Comment by jBarz** [comment]
> This only affects applications that don't call uv_loop_close()?

Yes.

**Comment by gireeshpunathil** [comment]
> This affects the test cases. Any source code using libuv would have to do the same.

shouldn't we document this? A libuv consumer may not expect the `loop` as holding onto persistent resources such as semaphores and message queues, instead may consider it as a data structure that is purely virtual memory resident with of course file descriptors (that gets cleaned up with process exit anyways) so may not bother to close the `loop`. What do you think?

**Comment by jBarz** [comment]
Thats a good idea. I will add that information to the documentation in this PR. Thanks for the suggestion.

**Comment by jBarz** [comment]
Could someone take a look at this? Unfortunately, the zos CI machine has run out of semaphores/msgqueues again. Once this change is integrated, no system resources should be leaked.

File: test/runner-unix.c [hunk 3]
```diff
@@ -39,10 +39,56 @@
...
+    }
+  }
+}
+
+
+static void abend_handler(int signal) {
+  release_system_resources();
+  raise(signal);
+}
+#endif
```
Comment (cjihrig) [inline test/runner-unix.c]: Could you indicate which `#ifdef` this matches up with.

File: test/runner-unix.c [hunk 4]
```diff
@@ -39,10 +39,56 @@
...
 
 
 /* Do platform-specific initialization. */
 int platform_init(int argc, char **argv) {
+  struct sigaction action;
```
Comment (cjihrig) [inline test/runner-unix.c]: Should this be behind an `#ifdef`?

File: test/runner-unix.c [hunk 5]
```diff
@@ -39,10 +39,56 @@
...
+      if (bufptr.msg.ipcqpcp.uid != getuid() || bufptr.msg.ipcqlspid != getpid()) {
+        if (foreignid == 0)
+          foreignid = bufptr.msg.ipcqmid;
+        else if (foreignid == bufptr.msg.ipcqmid) /* have we rotated to the top */
+          break;
+        continue;
+      }
+      msgctl(bufptr.msg.ipcqmid, IPC_RMID, NULL);
+    }
+    else if (memcmp(bufptr.common.ipcqtype, "ISEM", 4) == 0) {
```
Comment (cjihrig) [inline test/runner-unix.c]: Nit: `else if` on the previous line.

**Comment by jBarz** [comment]
Addressed all issues so far.

**Comment by gireeshpunathil** [comment]
LGTM.

**Comment by cjihrig** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/695/

**Comment by cjihrig** [comment]
Looks like zos tests are timing out instead now.

**Comment by jBarz** [comment]
I think the machine is out of resources because of the leakage from prior runs.
Someone will have to logon to the machine itself and manually clear them up.
There is a github work item to get me access to the machine https://github.com/nodejs/build/issues/1081
However, Michael is currently on vacation. If anyone else can get me access, that would be great.

**Comment by cjihrig** [comment]
Maybe @gibfahn can help?

**Comment by gibfahn** [comment]
Sounds like @mhdawson has given @jBarz access

**Comment by mhdawson** [comment]
@jBarz has access now.  Would still be good to get the PR landed to avoid having to periodically clean up.

**Comment by jBarz** [comment]
Could someone run a CI for this?
One test is expect to fail but https://github.com/libuv/libuv/pull/1711 fixed that.

**Comment by santigimeno** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/706/

**Comment by jBarz** [comment]
Could someone run this?
I fixed some of the timeout issues.

**Comment by mhdawson** [comment]
CI https://ci.nodejs.org/view/libuv/job/libuv-test-commit/741/

**Comment by jBarz** [comment]
Zos tests passed

**Comment by jBarz** [comment]
Hold off on merging this. There is one more thing I am investigating. Will have a fix by eod.

**Comment by jBarz** [comment]
Ok, i took care of some more edge cases. Would like for this to be run on more time.
After that, I will check on the machine to ensure no resources are left behind.

**Comment by mhdawson** [comment]
one more time: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/745/

**Comment by jBarz** [comment]
Only 1 failure. I have noticed this test timeout occasionally. This is not related to this PR.
```
not ok 99 - getaddrinfo_basic
# timeout
# Output from process `getaddrinfo_basic`:
# CEE5207E The signal SIGABRT was received.
# CEE5207E The signal SIGABRT was received.
# CEE5205S The signal SIGTERM was received.
```

I can reproduce this timeout about 1 in 100 tries on the same test machine.
I have increased the timeout from 5 seconds to 10 seconds for the getaddrinfo set of tests in a second commit. Lets have another try and see how it goes.

**Comment by mhdawson** [comment]
Started: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/749/

**Comment by jBarz** [comment]
Looks like it passes on z/OS
I checked the machine and it seems like no system resources are left behind by the tests.
So all looks good now.

**Comment by cjihrig** [comment]
This appears to break Windows - https://ci.nodejs.org/view/libuv/job/libuv-test-commit/769/

**Comment by jBarz** [comment]
I was able to add a cleanup script to the CI job with help from @mhdawson.
The tests are now cleaning up after themselves so we don't need this PR anymore.
